File: Backend/src/modules/food/dining/models/diningBooking.model.js
import mongoose from 'mongoose';

export const DINING_BOOKING_STATUSES = [
    'pending',
    'confirmed',
    'seated',
    'completed',
    'cancelled',
    'rejected',
    'no_show'
];

/** Statuses that still hold seats for a slot. */
export const DINING_ACTIVE_BOOKING_STATUSES = ['pending', 'confirmed', 'seated'];

const bookingHistorySchema = new mongoose.Schema(
    {
        status: { type: String, enum: DINING_BOOKING_STATUSES, required: true },
        note: { type: String, trim: true, default: '' },
        byRole: { type: String, trim: true, default: '' },
        byId: { type: mongoose.Schema.Types.ObjectId },
        at: { type: Date, default: Date.now }
    },
    { _id: false }
);

const bookedTableSchema = new mongoose.Schema(
    {
        tableId: { type: mongoose.Schema.Types.ObjectId, ref: 'FoodDiningTable' },
        name: { type: String, trim: true, default: '' },
        seats: { type: Number, default: 0 }
    },
    { _id: false }
);

const foodDiningBookingSchema = new mongoose.Schema(
    {
        bookingCode: { type: String, required: true, unique: true, trim: true, uppercase: true },
        userId: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true, index: true },
        restaurantId: {
            type: mongoose.Schema.Types.ObjectId,
            ref: 'FoodRestaurant',
            required: true,
            index: true
        },
        diningProfileId: { type: mongoose.Schema.Types.ObjectId, ref: 'FoodDiningProfile', index: true },

        // Day key at UTC midnight + "HH:mm" slot bounds; bookingAt is the exact instant (IST-derived).
        bookingDate: { type: Date, required: true, index: true },
        slotStart: { type: String, required: true, trim: true },
        slotEnd: { type: String, required: true, trim: true },
        bookingAt: { type: Date, required: true, index: true },

        guests: { type: Number, required: true, min: 1, max: 50 },
        tables: { type: [bookedTableSchema], default: [] },

        guestName: { type: String, required: true, trim: true, maxlength: 120 },
        guestPhone: { type: String, required: true, trim: true, maxlength: 20 },
        occasion: { type: String, trim: true, default: '', maxlength: 60 },
        specialRequest: { type: String, trim: true, default: '', maxlength: 500 },

        status: { type: String, enum: DINING_BOOKING_STATUSES, default: 'pending', index: true },
        statusHistory: { type: [bookingHistorySchema], default: [] },
        cancelledBy: { type: String, enum: ['user', 'restaurant', 'admin', ''], default: '' },
        cancelReason: { type: String, trim: true, default: '', maxlength: 300 },

        // Denormalized snapshot so history stays readable if the outlet changes later.
        restaurantName: { type: String, trim: true, default: '' },
        restaurantImage: { type: String, trim: true, default: '' },
        restaurantAddress: { type: String, trim: true, default: '' },

        rating: { type: Number, min: 1, max: 5 },
        review: { type: String, trim: true, default: '', maxlength: 500 },
        ratedAt: { type: Date }
    },
    {
        collection: 'food_dining_bookings',
        timestamps: true
    }
);

foodDiningBookingSchema.index({ restaurantId: 1, bookingDate: 1, status: 1 });
foodDiningBookingSchema.index({ restaurantId: 1, bookingDate: 1, slotStart: 1, status: 1 });
foodDiningBookingSchema.index({ userId: 1, createdAt: -1 });
foodDiningBookingSchema.index({ status: 1, bookingAt: 1 });

export const FoodDiningBooking = mongoose.model(
    'FoodDiningBooking',
    foodDiningBookingSchema,
    'food_dining_bookings'
);

File: foodrestaurent/lib/features/dining/domain/dining_booking_model.dart
class BookedTableModel {
  final String id;
  final String name;
  final int seats;

  const BookedTableModel({
    required this.id,
    required this.name,
    required this.seats,
  });

  factory BookedTableModel.fromJson(Map<String, dynamic> json) {
    return BookedTableModel(
      id: json['id']?.toString() ?? json['_id']?.toString() ?? '',
      name: json['name']?.toString() ?? '',
      seats: (json['seats'] as num?)?.toInt() ?? 0,
    );
  }

  Map<String, dynamic> toJson() => {
    'id': id,
    'name': name,
    'seats': seats,
  };
}

class DiningBookingModel {
  final String id;
  final String bookingCode;
  final String status;
  final String restaurantId;
  final String restaurantName;
  final String date;
  final String slotStart;
  final String slotEnd;
  final int guests;
  final List<BookedTableModel> tables;
  final String guestName;
  final String guestPhone;
  final String occasion;
  final String specialRequest;
  final String cancelledBy;
  final String cancelReason;
  final DateTime? createdAt;

  const DiningBookingModel({
    required this.id,
    required this.bookingCode,
    required this.status,
    required this.restaurantId,
    required this.restaurantName,
    required this.date,
    required this.slotStart,
    required this.slotEnd,
    required this.guests,
    required this.tables,
    required this.guestName,
    required this.guestPhone,
    required this.occasion,
    required this.specialRequest,
    required this.cancelledBy,
    required this.cancelReason,
    this.createdAt,
  });

  factory DiningBookingModel.fromJson(Map<String, dynamic> json) {
    final rawTables = json['tables'] as List? ?? [];
    return DiningBookingModel(
      id: json['id']?.toString() ?? json['_id']?.toString() ?? '',
      bookingCode: json['bookingCode']?.toString() ?? '',
      status: json['status']?.toString() ?? 'pending',
      restaurantId: json['restaurantId']?.toString() ?? '',
      restaurantName: json['restaurantName']?.toString() ?? '',
      date: json['date']?.toString() ?? '',
      slotStart: json['slotStart']?.toString() ?? '',
      slotEnd: json['slotEnd']?.toString() ?? '',
      guests: (json['guests'] as num?)?.toInt() ?? 1,
      tables: rawTables
          .map((t) => BookedTableModel.fromJson(Map<String, dynamic>.from(t as Map)))
          .toList(),
      guestName: json['guestName']?.toString() ?? '',
      guestPhone: json['guestPhone']?.toString() ?? '',
      occasion: json['occasion']?.toString() ?? '',
      specialRequest: json['specialRequest']?.toString() ?? '',
      cancelledBy: json['cancelledBy']?.toString() ?? '',
      cancelReason: json['cancelReason']?.toString() ?? '',
      createdAt: json['createdAt'] != null
          ? DateTime.tryParse(json['createdAt'].toString())
          : null,
    );
  }

  bool get isPending => status == 'pending';
  bool get isConfirmed => status == 'confirmed';
  bool get isSeated => status == 'seated';
  bool get isCompleted => status == 'completed';
  bool get isCancelled => status == 'cancelled';
  bool get isRejected => status == 'rejected';
  bool get isNoShow => status == 'no_show';
}
